### PR TITLE
[CUDA RNG] Allocate graph capture state on its capture stream

### PR DESCRIPTION
**Pull request by eee4017**
### Summary

This change allocates CUDA graph RNG state on the active capture stream while keeping it outside the graph execution pool.

### Motivation

The previous initializer switched to the default stream so the RNG seed and offset tensors would not enter the graph pool. That switch can expose an unrelated allocator event while capture is active:

```python
import torch

capture_stream = torch.cuda.Stream()
pending = torch.empty(1, device="cuda")
graph = torch.cuda.CUDAGraph()

with torch.cuda.stream(capture_stream):
    graph.capture_begin(capture_error_mode="thread_local")
    with torch.cuda.stream(torch.cuda.default_stream()):
        pending.record_stream(capture_stream)
        pending.untyped_storage().resize_(0)
    output = torch.rand(1, device="cuda")  # first RNG use
    graph.capture_end()
```

The first `torch.rand` creates two RNG state tensors. On the old path, their default-stream allocations can poll the pending event recorded on `capture_stream`, which fails with `cudaErrorCapturedEvent`. (See also https://github.com/pytorch/pytorch/issues/193982)

### Storage choice

RNG state could live in the graph execution pool, the default pool, or a shared RNG-only pool. The graph pool is natural but CUDA Graph Tree checkpoint restoration can invalidate state created after a checkpoint, while a shared RNG-only pool needs new lifetime and memory-reclamation bookkeeping. This change therefore keeps the existing default-pool placement, which uses normal allocator accounting and stays outside graph checkpoints, but allocates on the capture stream. A narrow allocator scope skips only the graph execution pool, preserves an explicit `MemPool`, and avoids new pool or checkpoint machinery.

### Remaining allocator limitation

The native caching allocator still has a broader edge case when a default-stream free records an event on a capturing stream and a later default-stream allocation polls it. That allocator-wide problem is tracked separately https://github.com/pytorch/pytorch/issues/193982.

RNG no longer enters that path after this change. A general fix would require broader target-stream event and capture-transition tracking, so it is outside this focused change.

### CODE REVIEW

File: aten/src/ATen/cuda/CUDAGeneratorImpl.cpp [hunk 1]
```diff
@@ -84,39 +84,55 @@ Generator createCUDAGenerator(DeviceIndex device_index) {
...
-  // mode may be Global (set by cudaStreamBeginCapture), which would block
-  // cudaMalloc even on a non-capturing stream.
-  c10::cuda::CUDAStreamCaptureModeGuard capture_mode_guard(
-      cudaStreamCaptureModeRelaxed);
-  c10::cuda::CUDAStreamGuard stream_guard(c10::cuda::getDefaultCUDAStream());
-
-  rng_state_seed_extragraph_ = at::empty({1}, options);
-  rng_state_offset_extragraph_ = at::empty({1}, options);
+  auto allocate_state = [&]() {
+    rng_state_seed_extragraph_ = at::empty({1}, options);
```
Comment (eqy) [inline aten/src/ATen/cuda/CUDAGeneratorImpl.cpp:103]: I assume two allocations is for alignment reasons?

Comment (eee4017) [reply aten/src/ATen/cuda/CUDAGeneratorImpl.cpp:103]: See the discussion here:

https://github.com/pytorch/pytorch/pull/173421#issuecomment-3888059130

**Comment by ngimel** [comment]
> CUDA Graph Tree checkpoint restoration can invalidate state created after a checkpoint

Can you give more details on how that would happen? I would really like graph-associated resources to live in graph pool(s), we are moving towards making that true for cublas workspaces and it's already true for cudnn, having a completely different cycle for rng only makes things confusing

**Comment by jbschlosser** [comment]
As an alternative to various CUDAGraph fixes on PyTorch's default RNG, note that I've been working on a set of new RNG APIs (currently beta in `torch/func/_random.py`) that are better behaved. In particular for CUDA, they accept explicit GPU-side RNG state to be used and interact intuitively with CUDAGraphs. These APIs don't have widespread usage yet, but if we suggest utilizing them over the old RNG when CUDAGraphability is desired, this would avoid the need for difficult-to-reason-about hacks when interacting with the old PyTorch RNG.

Let me explore possibilities for smooth transition to these APIs for the general user concerned with CUDAGraphability and report back; there are some feasible options that should require minimal user-code changes (albeit changes in generated numbers should be expected).

**Comment by eee4017** [comment]
@ngimel  The concern is the checkpointed pool used by CUDA Graph Trees. RNG state could live there, but the Tree would need to track it during checkpoint restoration. These tensors are owned by the generator, so they are currently invisible to the Tree’s list of live storages. Restoring the pool may therefore free and reconstruct their allocations while the generator still holds the old tensor storage. 

The reconstructed allocation must be reconnected to the generator’s tensor. We could expose each graph’s RNG storages to the Tree, record them with its checkpoint, and pass them through the existing storage-restoration path. However, their per-graph lifetime would need careful handling when switching between branches. This seems possible, but it is a broader and more complex change.

cc @eellison

**Comment by eee4017** [comment]
Thanks @jbschlosser. I have reviewed the stateless RNG work. I agree that explicit device-side keys are the cleaner long-term model for CUDA graphs. The RNG state will be a normal tensor supplied to the graph, rather than hidden generator-owned state. 

But I think we still need a PR for existing RNG path. Today many existing operators use the default generator, while the new API is beta and migration is not yet transparent. Do you planning to make a compatibility or compiler layer that replaces existing RNG calls such as `torch.rand` and dropout with the stateless implementation, without requiring users to use `key` through their model code?
